packages/timeline/src/browser/timeline-contribution.ts

```diff
@@ -0,0 +1,108 @@
+ ********************************************************************************/
+
+import { injectable, inject } from 'inversify';
+import {
+    FrontendApplicationContribution,
+    FrontendApplication,
+    ViewContainer,
+    WidgetManager,
+    Widget,
+    ApplicationShell,
+    Navigatable
+} from '@theia/core/lib/browser';
+import { FileNavigatorContribution } from '@theia/navigator/lib/browser/navigator-contribution';
+import { EXPLORER_VIEW_CONTAINER_ID } from '@theia/navigator/lib/browser';
+import { TimelineWidget } from './timeline-widget';
+import { TimelineService } from './timeline-service';
+import { Command, CommandRegistry } from '@theia/core/lib/common';
+import { TabBarToolbarRegistry } from '@theia/core/lib/browser/shell/tab-bar-toolbar';
+
+@injectable()
+export class TimelineContribution implements FrontendApplicationContribution {
+
+    @inject(FileNavigatorContribution)
+    protected readonly explorer: FileNavigatorContribution;
+    @inject(WidgetManager)
+    protected readonly widgetManager: WidgetManager;
+    @inject(TimelineService)
+    protected readonly timelineService: TimelineService;
+    @inject(CommandRegistry)
+    protected readonly commandRegistry: CommandRegistry;
+    @inject(TabBarToolbarRegistry)
+    protected readonly tabBarToolbar: TabBarToolbarRegistry;
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    public static readonly LOAD_MORE_COMMAND: Command = {
+        id: 'timeline-load-more'
+    };
+
+    async onDidInitializeLayout?(app: FrontendApplication): Promise<void> {
+        const explorer = await this.widgetManager.getWidget(EXPLORER_VIEW_CONTAINER_ID);
+        let timeline: TimelineWidget;
+        this.timelineService.onDidChangeProviders( async event => {
+            if (explorer instanceof ViewContainer) {
+                if (event.added && event.added.length > 0 && explorer.getTrackableWidgets().indexOf(timeline) === -1) {
+                    timeline = await this.widgetManager.getOrCreateWidget(TimelineWidget.ID);
+                    explorer.addWidget(timeline, { initiallyCollapsed: true });
+                } else if (event.removed && this.timelineService.getSources().length === 0) {
+                    timeline.close();
+                }
+            }
+        });
+        const toolbarItem = {
+            id: 'timeline-refresh-toolbar-item',
+            command: 'timeline-refresh',
+            tooltip: 'Refresh',
+            icon: 'fa fa-refresh'
+        };
+        this.commandRegistry.registerCommand({ id: toolbarItem.command }, {
+            execute: widget => this.checkWidget(widget, () => {
+                if (timeline) {
+                    timeline.refreshList();
+                }
+            }),
+            isEnabled: widget => this.checkWidget(widget, () => true),
+            isVisible: widget => this.checkWidget(widget, () => true)
+        });
+        let navigable: Navigatable;
```

Comment:
please don't cache, use the shell as a source of truth, by always reading from `shell.currentWidget`

Reply:
When a timeline item is clicked `shell.currentWidget` returns the `timelineWidget` because it was activated after the mouse click, but we need the current `Navigable` widget.

Reply:
done